File: Python3/프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
import collections
import sys

def solution(edges):
    sys.setrecursionlimit(10000000)
    graph = collections.defaultdict(list)
    
    # 그래프, 노드수 정의
    nodes = set()
    outdegrees = collections.defaultdict(int)
    indegrees = collections.defaultdict(int)
    for edge in edges:
        graph[edge[0]].append(edge[1])
        nodes.add(edge[0])
        nodes.add(edge[1])
        outdegrees[edge[0]] += 1 # ✅고친부분✅
        indegrees[edge[1]] += 1 # ✅고친부분✅
        
    # 시작노드 찾기 ✅고친부분✅
    start_node = 0
    for node in nodes:
        if node not in indegrees and outdegrees[node] >= 2:
            start_node = node
            break
        
    print(start_node)
      
    # DFS 함수
    def dfs(start, cur, visited):
        # print(cur)
        if len(graph[cur]) == 2:
            # print("팔자")
            return 3
        if len(graph[cur]) == 0:
            # print("막대")
            return 2
        for next in graph[cur]:
            if next not in visited:
                visited.add(next)
                return dfs(start, next, visited)
        return 1
    
    # 시작지점에서부터 노드들 순회 돌기
    result = [start_node, 0, 0, 0]
    for cur in graph[start_node]:
        # print("----------cur-------------")
        idx = dfs(cur, cur, set())
        if idx != None:
            result[idx] += 1
        
    return result

File: Swift/프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.swift
import Foundation

func solution(_ edges:[[Int]]) -> [Int] {
    
    // 노드간 관계성 정리
    var graph = [Int:[Int]]()
    var nodes = Set<Int>()
    var indegree = Set<Int>()
    var outdegree = [Int: Int]()
    for edge in edges {
        graph[edge[0], default: []].append(edge[1])
        nodes.insert(edge[0])
        nodes.insert(edge[1])
        indegree.insert(edge[1])
        outdegree[edge[0], default: 0] += 1
    }
    
    // 시작 노드 찾기
    var start = 0
    for node in nodes {
        if !indegree.contains(node) && outdegree[node]! >= 2 {
            start = node
        }
    }
    
    var result = [start,0,0,0]
    
    // 노드 순회하기
    for node in graph[start]! {
        var queue = [Int]()
        var visited = Set<Int>()
        queue.append(node)
        
        // flag 1: 도넛, 2: 막대, 3: 8자
        var flag = 2
        
        while !queue.isEmpty {
            var temp = queue.removeFirst()
            guard let next = graph[temp] else { break }
            
            // 8자인경우
            if next.count == 2 {
                flag = 3
                break
            }
            
            if visited.contains(next[0]) {
                flag = 1
                break
            }
            
            if next.count != 0 {
                queue.append(next[0])
                visited.insert(next[0])
            }
        }
        
        result[flag] += 1
    }
    
    return result
}